### PR TITLE
Add an ENUM setting for disabling rendering "intense" text as bold

### LINKED ISSUE #10576: Add a setting for disabling "intense is bold"
A follow up to #10498

As discussed at length in other threads, esp #109.

Most recently, from @j4james in #10498:

> I mean choose if you want SGR 1 to select a bold font face or not (basically a way to disable this PR). I suppose you could potentially implement that by allowing users to choose which font weight is bold, and expect them to select the normal font weight, but that seems a bit obscure. For example, in XTerm you can just uncheck the "Bold Fonts" menu, and in Konsole you uncheck the option "Draw intense colors in bold font".
> 
> There are a whole bunch of other variations people may want to configure, e.g. disabling bold-as-bright, or only enabling a bold font when used with the extended colors (i.e. the aix bright colors and the ITU 256/RGB colors). Personally I think the latter is the best default config, since it gives you the standard behaviour for the original 8 SGR colors, while still allowing bold fonts when used with extended colors, but I'd be happy with a simple option to turn this off.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

This adds a new setting `intenseTextStyle`. It's a per-appearance, control setting, defaulting to `"all"`.
* When set to `"all"` or `["bold", "bright"]`, then we'll render text as both **bold** and bright (1.10 behavior)
* When set to `"bold"`, `["bold"]`, we'll render text formatted with `^[[1m` as **bold**, but not bright
* When set to `"bright"`, `["bright"]`, we'll render text formatted with `^[[1m` as bright, but not bold. This is the pre 1.10 behavior
* When set to `"none"`, we won't do anything special for it at all. 

## references
* I last did this in #10648. This time it's an enum, so we can add bright in the future. It's got positive wording this time.
* ~We will want to add `"bright"` as a value in the future, to disable the auto intense->bright conversion.~ I just did that now.
* #5682 is related

## PR Checklist
* [x] Closes #10576 
* [x] I seriously don't think we have an issue for "disable intense is bright", but I'm not crazy, people wanted that, right? https://github.com/microsoft/terminal/issues/2916#issuecomment-544880423 was the closest
* [x] I work here
* [x] Tests added/passed
* [x] https://github.com/MicrosoftDocs/terminal/pull/381

## Validation Steps Performed

image

Yea that works. Printed some bold text, toggled it on, the text was no longer bold. hooray.

### EDIT, 10 Aug

```json
"intenseTextStyle": "none",
"intenseTextStyle": "bold",
"intenseTextStyle": "bright",
"intenseTextStyle": "all",
"intenseTextStyle": ["bold", "bright"],
```

all work now. Repro script:
```sh
printf "\e[1m[bold]\e[m[normal]\e[34m[blue]\e[1m[bold blue]\e[m\n"
```

### CODE REVIEW

File: src/cascadia/TerminalSettingsModel/IAppearanceConfig.idl [hunk 1]
```diff
@@ -22,6 +22,14 @@ namespace Microsoft.Terminal.Settings.Model
         Vertical_Bottom = 0x20
     };
 
+    [flags]
+    enum IntenseStyle
+    {
+        Bold = 0x1,
+        // Bright = 0x2,
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/IAppearanceConfig.idl:29]: might as well leave this one uncommented. after all, it's only accessible if you add it to the enum mapper

File: src/renderer/dx/DxRenderer.hpp [hunk 2]
```diff
@@ -256,6 +257,7 @@ namespace Microsoft::Console::Render
         D2D1_TEXT_ANTIALIAS_MODE _antialiasingMode;
 
         float _defaultTextBackgroundOpacity;
+        bool _intenseIsBold;
```
Comment (DHowett) [inline src/renderer/dx/DxRenderer.hpp:260]: nit: you can initialize this in the class body if you want.

**Review by DHowett — APPROVED** [review]

**Review by leonMSFT — APPROVED** [review]

**Comment by j4james** [comment]
> When set to "none", we won't. (the pre-1.10 behavior)

This sounds wrong to me. If the long term plan is to allow the choice of "none", "bright", "bold", and "all", then the pre-1.10 behavior is "bright" rather than "none".

**Comment by zadjii-msft** [comment]
> This sounds wrong to me. If the long term plan is to allow the choice of "none", "bright", "bold", and "all", then the pre-1.10 behavior is "bright" rather than "none".

Sorry yea, that is a little confusing. Right now, there's only two values that the settings will actually accept - `bold` and `none`. `none` will still do the bright thing that we've always done right now, until we come through and actually add support for the `bright` setting. Then `none` will be (not bright, not bold). 

I'm not sure I can fiddle with the enum parsing enough in this PR to make it so the only options are some set of `bold`, `bright`, and `all`. Maybe this PR doesn't make any sense at all without also doing the bright setting (though that's harder 😕)

File: src/cascadia/TerminalControl/IControlAppearance.idl [hunk 3]
```diff
@@ -11,6 +11,8 @@ namespace Microsoft.Terminal.Control
         Windows.UI.Xaml.Media.Stretch BackgroundImageStretchMode;
         Windows.UI.Xaml.HorizontalAlignment BackgroundImageHorizontalAlignment;
         Windows.UI.Xaml.VerticalAlignment BackgroundImageVerticalAlignment;
+        Boolean IntenseIsBold;
+        // Boolean IntenseIsBright;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/IControlAppearance.idl:15]: Why is `IntenseStyle` declared in the `Microsoft.Terminal.Settings.Model` instead of the `Microsoft.Terminal.Control` namespace? I guess you need to convert the enum into two bools at some point, but it's just weird to see two super similar settings in one place imo.

File: src/cascadia/TerminalSettingsModel/AppearanceConfig.h [hunk 4]
```diff
@@ -52,6 +52,7 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
         INHERITABLE_SETTING(Model::IAppearanceConfig, bool, RetroTerminalEffect, false);
         INHERITABLE_SETTING(Model::IAppearanceConfig, hstring, PixelShaderPath, L"");
+        INHERITABLE_SETTING(Model::IAppearanceConfig, Microsoft::Terminal::Settings::Model::IntenseStyle, IntenseTextStyle, Microsoft::Terminal::Settings::Model::IntenseStyle::All);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/AppearanceConfig.h:55]:
```suggestion
        INHERITABLE_SETTING(Model::IAppearanceConfig, IntenseStyle, IntenseTextStyle, IntenseStyle::All);
```

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
Don't forget to update the schema and the docs! (only reason I'm blocking)

Everything else is a suggestion.

**Comment by j4james** [comment]
> Right now, there's only two values that the settings will actually accept - `bold` and `none`.

Maybe I'm not understanding how these mappings work, but if those are the actual values that are going to be saved in the settings file when a user toggles the "Render intense text as bold" option, then surely that setting is guaranteed to be incorrect at some point in the future?

For example, if a user turns this option off (so the setting becomes _none_), `SGR 1` will still show as bright for now, but will eventually stop working altogether. Even when it's turned on, it's going to change from bold+bright to just bold. We're almost certain to receive a whole lot of bug reports when that happens.

If it can't be avoided, so be it. I just wanted to be sure you were aware of the problem. And if I've just misunderstood how it's going to work, sorry for the noise.

**Comment by skyline75489** [comment]
>  `none` will still do the bright thing that we've always done right now, until we come through and actually add support for the `bright` setting. Then `none` will be will be (not bright, not bold).

Yea I agree with @j4james , this change of behavior will result in huge amount of bug reports.

>  this PR doesn't make any sense at all without also doing the bright setting

Adding the bright setting is likely the correct solution here 😦

**Comment by zadjii-msft** [comment]
shh bot, I'm coming back to this

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 5]
```diff
@@ -600,6 +601,7 @@ namespace winrt::Microsoft::Terminal::Control::implementation
 
         _renderEngine->SetForceFullRepaintRendering(_settings.ForceFullRepaintRendering());
         _renderEngine->SetSoftwareRendering(_settings.SoftwareRendering());
+
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:604]:
```suggestion
```
nit: unnecessary change

**Review by carlos-zamora — COMMENTED** [review]
Don't forget to update the schema and the docs! (only reason I'm blocking)

Everything else is a suggestion.

**Review by carlos-zamora — APPROVED** [review]

**Comment by DHowett** [comment]
build failures :(

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 6]
```diff
@@ -1186,4 +1186,28 @@
...
+    <value>Bold</value>
+    <comment>An option to choose from for the "intense text format" setting. When selected, "intense" text will be rendered as bold text</comment>
+  </data>
+  <data name="Appearance_IntenseTextStyleBright.Content" xml:space="preserve">
+    <value>Bright</value>
+    <comment>An option to choose from for the "intense text format" setting. When selected, "intense" text will be rendered in a brighter color</comment>
+  </data>
+  <data name="Appearance_IntenseTextStyleAll.Content" xml:space="preserve">
+    <value>Both</value>
+    <comment>An option to choose from for the "intense text format" setting. When selected, "intense" text will be rendered as both bold text and in a brighter color</comment>
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1211]: Worried about the wording here. The names are pretty short, and people may not know what "Intense Text Format" means. Should we do something more like `Display "intense" text in...` `...the normal font` `a bold font` `...bright colors` `...both bold font and bright colors`?

Should we get more technical and mention SGR 1?

File: src/cascadia/TerminalControl/IControlAppearance.idl [hunk 7]
```diff
@@ -11,6 +11,8 @@ namespace Microsoft.Terminal.Control
         Windows.UI.Xaml.Media.Stretch BackgroundImageStretchMode;
         Windows.UI.Xaml.HorizontalAlignment BackgroundImageHorizontalAlignment;
         Windows.UI.Xaml.VerticalAlignment BackgroundImageVerticalAlignment;
+        Boolean IntenseIsBold;
+        // IntenseIsBright is in Core Appearance
```
Comment (DHowett) [inline src/cascadia/TerminalControl/IControlAppearance.idl:15]: technically both should be there, since eventually we'll push down the renderer right?

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 6 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1211]: @cinnamon-msft what do you think? this is a button for users who "know what they're doing"

File: src/cascadia/TerminalControl/IControlAppearance.idl (hunk 7 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/IControlAppearance.idl:15]: IMO eventually is not _today_, and I'd rather be self-consistent with the state of the world and not what we expect the world to be (which may lead to confusion)

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1211]: I'm gonna guess that .1% of our users will know what SGR 1 is. I be 1% of the folks that upvoted #109 in the first place do - people just want bold. 

I'm not a wordsmith, [I'm just a lowly engineer](https://memegenerator.net/img/instances/400x/74828021.jpg)

Comment (j4james) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1211]: As a point of comparison, Mintty has this option as a dropdown list with the label "Show bold", and with the choices "as font", "as color", or "as font & as color". XTerm just as the option to toggle the bold fonts (at least in the UI) with a checked menu item labelled "Bold Fonts". And Gnome Terminal just as the option to toggle the bright colors with a checkbox labelled "Show bold text in bright colors".

My recommendation would be something similar to what Mintty has, but with a little more detail, e.g.:
- Show bold with brighter colors
- Show bold with a heavier font weight
- Show bold with both font weight and color

I think most people that are going to want to mess with this know that it's the "bold" functionality that they're tweaking. And what we're trying to differentiate is whether it affects the font or the color (or both).

**Review by skyline75489 — APPROVED** [review]
:shipit:

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 6 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1211]: @j4james I appreciate that! Thanks. We're gonna merge this as is to get the feature into selfhost/test builds and work on the wording from there. :)

**Review by DHowett — APPROVED** [review]

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1211]: (we're gonna work on the wording in post)
